### DISCUSSION TITLE
Label API: Get the date when a label was added to an issue.

### DISCUSSION THREAD

**Discussion by dereksmart** [community/community#24630] [Apps, API and Webhooks]
<p>I am looking for a way to get the date/time a label was added to an issue in order to determine how long something has been labeled “needs review” (for example). It would be great if this endpoint returned the date of the labels added. </p>
<pre><code>GET /repos/:owner/:repo/issues/:number/labels
</code></pre>
<p>Or is there any other way to do this through the API? </p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/dereksmart">@dereksmart</a>,</p>
<p>Thanks for reaching out! There is a way in v4 using createdAt like in the query below, maybe someone else has found another way but I hope this helps!</p>
<pre><code>query {
  resource(url:"https://github.com/atom/atom/pull/10999") {
    ... on PullRequest {
      labels(first:10) {
        edges {
          label: node {
            name
            createdAt
          }
        }
      }
    }
  }
} 
</code></pre>

**Comment by dereksmart** [comment]
<p><a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> Thanks for the reply, but <code>createdAt</code> seems to give the date in which the label was created. I am looking for the date the label was added to the PR/issue. </p>

**Comment by nickvanw** [comment]
<p>Hey <a class="mention" href="/dereksmart">@dereksmart</a>!</p>
<p>I believe this is possible by using the Timeline APIs (<a href="https://developer.github.com/v3/issues/timeline/" rel="nofollow noopener">v3</a>, <a href="https://developer.github.com/v4/object/labeledevent/" rel="nofollow noopener">v4</a> from the timeline connection). </p>
<p>Using the REST API, I’m able to query:</p>
<pre><code>curl -H "Accept: application/vnd.github.mockingbird-preview" https://api.github.com/repos/atom/atom/issues/10999/timeline
</code></pre>
<p>This gives me a list of events, including “labeled” events that have a “created_at” property that should indicate when the label was added.</p>
<p>Let me know if this helps!</p>

**Comment by dereksmart** [comment]
<p>Hi <a class="mention" href="/nickvanw">@nickvanw</a> - thanks for the reply. </p>
<p>Yes, Timeline seems to be the best way to go here. </p>
<p>Without having to send multiple requests (one per issue/PR), this is the best I’ve come up with so far for getting multiple issue timelines in a single request: </p>
<pre><code>query{
  repository(owner: "ownerName", name: "repoName") {
    pullRequests(last: 100, states: OPEN, labels:"[Status] Needs Review" ) {
      totalCount
      edges {
        node {
          title
          url
          number
          timeline(last: 100) {
            totalCount
            edges {
              node {
                __typename
                ... on LabeledEvent {
                  createdAt
                  label {
                    name
                  }
                }
                ... on UnlabeledEvent {
                  createdAt
                  label {
                    name
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>And then whatever is consuming the response will need to do some maual filtering and calculations to see if the label is still applied, and get the date from the last applied instance. </p>
<p>Suggestions welcome on how I could simplify this! </p>

**Comment by nickvanw** [comment, marked as answer]
<p>Hey <a class="mention" href="/dereksmart">@dereksmart</a>!</p>
<p>That looks good to me! If I have one suggestion, it might be to use the pullRequest.timelineItems connection that’s currently in preview: <a href="https://developer.github.com/v4/previews/#issues-preview" rel="nofollow noopener">https://developer.github.com/v4/previews/#issues-preview</a></p>
<p>This connection allows filtering the timeline down to a specific list of items, so you could do something like this:</p>
<pre><code>query{
  repository(owner: "owner", name: "name") {
    pullRequests(last: 10, states: OPEN) {
      totalCount
      edges {
        node {
          title
          url
          number
          timelineItems(last: 100, itemTypes: [LABELED_EVENT, UNLABELED_EVENT]) {
            totalCount
            edges {
              node {
                __typename
                ... on LabeledEvent {
                  createdAt
                  label {
                    name
                  }
                }
                ... on UnlabeledEvent {
                  createdAt
                  label {
                    name
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>Some manual filtering will still be required, but this means that you’ll get less noise in your query. </p>

**Comment by dereksmart** [comment]
<p>timelineItems is a beautiful suggestion. Thank you! </p>

**Comment by ryparker** [comment]
<p>Best workaround so far nickvanw@, thank you.</p>
<p>I hope they plan to make this easier to query down the road.</p>
